File: lessons/dsa/stack-and-last-in-order/examples/last_in_order_test.go
package examples

import (
	"slices"
	"strings"
	"testing"
)

var work = []string{"render page", "lay out list", "measure row"}

func reversed(items []string) []string {
	out := slices.Clone(items)
	slices.Reverse(out)
	return out
}

// The defining property, over many inputs rather than one. A structure that
// failed this would not be a stack whatever its methods were named.
func TestWhatComesOutIsWhatWentInReversed(t *testing.T) {
	inputs := [][]string{
		{},
		{"only"},
		work,
		{"a", "b", "c", "d", "e", "f"},
		{"same", "same", "same"},
	}
	for _, items := range inputs {
		if got := Drain(items); !slices.Equal(got, reversed(items)) {
			t.Fatalf("draining %v gave %v", items, got)
		}
	}
}

func TestTheLessonOrder(t *testing.T) {
	want := []string{"measure row", "lay out list", "render page"}
	if got := Drain(work); !slices.Equal(got, want) {
		t.Fatalf("expected %v, got %v", want, got)
	}
}

func TestDrainingTwiceRestoresTheOriginalOrder(t *testing.T) {
	if got := Drain(Drain(work)); !slices.Equal(got, work) {
		t.Fatalf("expected %v, got %v", work, got)
	}
}

func TestBalancedNestingIsAccepted(t *testing.T) {
	for _, text := range []string{"", "()", "([{}])", "a(b)c[d]e", "(())()"} {
		if !IsBalanced(text) {
			t.Fatalf("expected %q to be balanced", text)
		}
	}
}

func TestUnbalancedNestingIsRejected(t *testing.T) {
	// The last case is the one a counter of opens and closes cannot detect: the
	// counts match and the nesting is still wrong.
	for _, text := range []string{")", "())", "(", "([)", "([)]"} {
		if IsBalanced(text) {
			t.Fatalf("expected %q to be rejected", text)
		}
	}
	if strings.Count("([)]", "(")+strings.Count("([)]", "[") != 2 {
		t.Fatal("expected two openers in the counting counterexample")
	}
}

// A thousand pairs in sequence never need more than one slot; ten nested need
// ten. This is why recursion depth is the thing to reason about.
func TestDepthGrowsWithNestingNotWithLength(t *testing.T) {
	if got := DeepestNesting(strings.Repeat("()", 1000)); got != 1 {
		t.Fatalf("expected depth 1, got %d", got)
	}
	nested := strings.Repeat("(", 10) + strings.Repeat(")", 10)
	if got := DeepestNesting(nested); got != 10 {
		t.Fatalf("expected depth 10, got %d", got)
	}
}

func TestTextWithoutBracketsIsBalancedAndFlat(t *testing.T) {
	if !IsBalanced("no brackets here") || DeepestNesting("no brackets here") != 0 {
		t.Fatal("expected plain text to be balanced and flat")
	}
}

File: lessons/dsa/stack-and-last-in-order/examples/last-in-order.test.ts
import { expect, test } from "vitest";

import { deepestNesting, drain, isBalanced } from "./last-in-order.js";

const WORK = ["render page", "lay out list", "measure row"];

test("what comes out is what went in, reversed", () => {
  // The defining property, over many inputs rather than one. A structure that
  // failed this would not be a stack whatever its methods were named.
  const wrong = [
    [],
    ["only"],
    WORK,
    [..."abcdef"],
    ["same", "same", "same"],
  ].filter(
    (items) =>
      JSON.stringify(drain(items)) !== JSON.stringify([...items].reverse()),
  );

  expect(wrong).toEqual([]);
});

test("the lesson's order", () => {
  expect(drain(WORK)).toEqual(["measure row", "lay out list", "render page"]);
});

test("draining twice restores the original order", () => {
  expect(drain(drain(WORK))).toEqual(WORK);
});

test("balanced nesting is accepted", () => {
  const rejected = ["", "()", "([{}])", "a(b)c[d]e", "(())()"].filter(
    (text) => !isBalanced(text),
  );

  expect(rejected).toEqual([]);
});

test("a closer with nothing open is rejected", () => {
  expect([isBalanced(")"), isBalanced("())")]).toEqual([false, false]);
});

test("something left open is rejected", () => {
  expect([isBalanced("("), isBalanced("([)")]).toEqual([false, false]);
});

test("the wrong closer is rejected", () => {
  // The case a counter of opens and closes cannot detect: the counts match and
  // the nesting is still wrong.
  const opens = [..."([)]"].filter((c) => "([".includes(c)).length;
  const closes = [..."([)]"].filter((c) => ")]".includes(c)).length;

  expect({ balanced: isBalanced("([)]"), closes, opens }).toEqual({
    balanced: false,
    closes: 2,
    opens: 2,
  });
});

test("depth grows with nesting, not with length", () => {
  // A thousand pairs in sequence never need more than one slot; ten nested need
  // ten. This is why recursion depth is the thing to reason about.
  expect({
    nested: deepestNesting("(".repeat(10) + ")".repeat(10)),
    sequential: deepestNesting("()".repeat(1000)),
  }).toEqual({ nested: 10, sequential: 1 });
});

test("text without brackets is balanced and flat", () => {
  expect({
    balanced: isBalanced("no brackets here"),
    depth: deepestNesting("no brackets here"),
  }).toEqual({ balanced: true, depth: 0 });
});
